pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+package ftp
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/textproto"
+	"os"
+	"path"
+	"time"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(filepath string) (ftpResponse, error) {
+	return c.ServerConn.Retr(filepath)
+}
+
+func (c *Conn) RetrFrom(filepath string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(filepath, offset)
+}
+
+var (
+	errEmptyFilename            = errors.New("filename cannot be empty")
+	errEmptyPath                = errors.New("file/directory path cannot be empty")
+	errEmptyDirectory           = errors.New("directory name cannot be empty")
+	errInvalidArg               = errors.New("invalid filename/directory")
+	transferCompleteError       = &textproto.Error{Code: 226, Msg: "Transfer complete."}
+	directoryAlreadyExistsError = &textproto.Error{Code: 550, Msg: "Create directory operation failed."}
```

Comment:
I'm not even sure we need these in fact

See https://github.com/gofr-dev/gofr/pull/871#discussion_r1699068172